src/process/io/s3/S3Source.h

```diff
@@ -0,0 +1,222 @@
+/**
+ * Copyright (c) 2019 DKFZ - ODCF
+ *
+ * Distributed under the MIT License (license terms are at https://github.com/dkfz-odcf/FastqIndEx/blob/master/LICENSE.txt).
+ */
+
+#ifndef FASTQINDEX_S3Source_H
+#define FASTQINDEX_S3Source_H
+
+#include "S3Config.h"
+#include "S3GetObjectProcessWrapper.h"
+#include "process/io/Source.h"
+#include "process/io/StreamSource.h"
+#include "FQIS3Client.h"
+
+#include <aws/core/Aws.h>
+#include <aws/core/auth/AWSCredentialsProvider.h>
+#include <aws/s3/S3Client.h>
+#include <aws/s3/model/GetObjectRequest.h>
+#include <fstream>
+#include <iostream>
+#include <semaphore.h>
+
+/**
+ * This class enables you to get data from an S3 bucket. The way, how we retrieve data, might seem odd. Unfortunately,
+ * we experienced quite a number of issues with the Amazon AWS SDK and eventually decided to do everything like this.
+ * - The SDK does not allow to work directly with the data stream. Therefore, we decided to download the data to
+ *   a named pipe in the tmp folder and access this with a StreamSource
+ * - Due to this, it is not possible to seek within an S3 file stream! Fortunately, the StreamSource has some built-in
+ *   buffer to deal with this.
+ * - It is not possible to abort a running request! What we did first was download with GetObjectAsync. However this
+ *   lead to errors, when we closed the pipe or tried to close the stream. Worst case: The application crashed, best
+ *   case: It didn't crash but always showed an error (SIGPIPE / SIGABRT). Personally, my trust in application is not
+ *   very high, if it shows me such types of errors.
+ *   Eventually we moved the download to a separate process. It works but has no real error handling by now. But: It
+ *   swallows any bad error codes and messages like a broken pipe (which is not nice but intentional in this case).
+ * Unfortunately, we are not able to really know, how much data we need for extraction and we need the possibility to
```

Comment:
With this implementation:

Can it happen that data is incomplete and corrupt output is created without notice? This would be the absolute worst case and ought not to happen -- then even the error messages would be better!

Please comment on this and maybe (ideally) we'd need tests to ensure things go right!